### PR TITLE
[NEARLY READY] CO-OP MECHA IS BACK my friends

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

COOP MECHA! TWO PEOPLE RIDE THIS BASTARD! ONE FOR MOVEMENT AND ONE FOR GUNS. BOTH SIDES GET A SPECIAL MECH ABILITY THEY MUST TIME TO DEAL EXCELLENT DAMAGE!

LATE GAME TECHWEBS UNLOCK THIS (AROUND THE SAME TIME AS PHAZON BUT WITH DIFFERENT BRANCHES OF TECH AS PREREQUISITE

Movement mech gets a mechanical leap that knocks people down for a long time, and violently throws mechs away from them!

Weapons mech launches rockets! They take a long time to land but deal immense amount of damage. This will surely miss unless you had some kind of ability to knock someone down for a long time...!

## Why It's Good For The Game

A REALLY STRONG MECH CHOICE IF YOU CAN TRUST THE PERSON IN THE OTHER SEAT! MUCH COOPERATION REQUIRED! 

# PACIFIC RIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIM WE GOT SPRITES FINALLY WOOP WOOP

## Changelog
:cl: Yeeye sprited this masterpiece
add: COOP MECHA NOW BUILDABLE (requires very late game tech, so get 'webbing!)
/:cl:

### CODE REVIEW

**Comment by wesoda25** [comment]
I am not a bitch please do not call me a bitch

File: code/game/mecha/combat/nerchen.dm [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	deflect_chance = 25
+	armor = list("melee" = 30, "bullet" = 30, "laser" = 30, "energy" = 30, "bomb" = 30, "bio" = 0, "rad" = 50, "fire" = 100, "acid" = 100)
+	max_temperature = 25000
+	infra_luminosity = 3
+	wreckage = /obj/structure/mecha_wreckage/nerchen
+	melee_can_hit = FALSE
+	add_req_access = 1
+	internal_damage_threshold = 25
+	var/leaping = TRUE
+	var/obj/mecha/combat/chen/chen //keeps the other internal mech in check
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]: add dmdoc

**Comment by tralezab** [comment]
by the order of the jarl you will compile

File: code/game/mecha/combat/nerchen.dm [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	if(occupant)
+		newicon += "closed"
+	else
+		newicon += "open"
+	newicon += "-"
+	if(chen.occupant)
+		newicon += "closed"
+	else
+		newicon += "open"
+	icon_state = newicon
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
	icon_state = "[initial(icon_state)]-[occupant ? "closed" : "open"]-[chen.occupant ? "closed" : "open"]"
```

File: code/game/mecha/combat/nerchen.dm [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+/obj/mecha/combat/nerchen/throw_impact(atom/hit_atom, datum/thrownthing/throwingdatum)
+	cockpit_icon()
+	..()
+
+/obj/mecha/combat/nerchen/mmi_move_inside(obj/item/mmi/mmi_as_oc, mob/user) //no cheese
+	to_chat(user, "<span class='warning'>There doesn't seem to be any way to interface with the mech!</span>")
+	return FALSE
+
+/obj/mecha/combat/nerchen/proc/cockpit_icon() //icon proc, will also reset
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]: dmdoc

File: code/game/mecha/combat/nerchen.dm [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+/obj/mecha/combat/chen
+	desc = "The second seat for the Nerchen."
+	name = "Gunner Seat"
+	icon_state = "phazon"
+	dir_in = 2 //Facing South. not sure if this one matters?
+	deflect_chance = 30
+	armor = list("melee" = 30, "bullet" = 30, "laser" = 30, "energy" = 30, "bomb" = 30, "bio" = 0, "rad" = 50, "fire" = 100, "acid" = 100)
+	max_temperature = 25000
+	infra_luminosity = 3
+	wreckage = null
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]: double check this doesnt cause runtime

File: code/game/mecha/combat/nerchen.dm [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+	add_req_access = 1
+	internal_damage_threshold = 25
+	force = 15
+	max_equip = 3
+	var/obj/mecha/combat/nerchen/ner
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]: dmdoc

File: code/game/mecha/combat/nerchen.dm [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	if(dna_lock)
+		var/passed = FALSE
+		if(user.has_dna())
+			var/mob/living/carbon/C = user
+			if(C.dna.unique_enzymes==dna_lock)
+				passed = TRUE
+		if (!passed)
+			to_chat(user, "<span class='warning'>Access denied. [name] is secured with a DNA lock.</span>")
+			log_message("Permission denied (DNA LOCK).", LOG_MECHA)
+			return
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
	if(dna_lock && user.has_dna())
		var/mob/living/carbon/C = user
		if(C.dna.unique_enzymes != dna_lock)
			to_chat(user, "<span class='warning'>Access denied. [name] is secured with a DNA lock.</span>")
			log_message("Permission denied (DNA LOCK).", LOG_MECHA)
			return
```

File: code/game/mecha/combat/nerchen.dm [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+/obj/mecha/combat/chen/click_action(atom/target,mob/user,params, chencommand = FALSE)
+	ner.click_action(target, user, params, chencommand = TRUE)
+
+#define LEAP_COOLDOWN 30 SECONDS
+
+//action buttons
+/datum/action/innate/mecha/leap
+	name = "Mecha Leap"
+	button_icon_state = "mech_missile"
+	var/cooldowns = 0 //i know we have a cooldown action but other mecha action buttons have not used it
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]: dmdoc

File: code/game/mecha/combat/nerchen.dm [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+/datum/action/innate/mecha/leap/Activate(forced_state = null)
+	if(!owner || !chassis || chassis.occupant != owner)
+		return
+	if(!isnull(forced_state))
+		chassis.leg_overload_mode = forced_state
+	else
+		chassis.leg_overload_mode = !chassis.leg_overload_mode
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
	chassis.leg_overload_mode = !isnull(forced_state) ? forced_state : !chassis.leg_overload_mode
```

File: code/game/mecha/combat/nerchen.dm [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+	if(chassis.leg_overload_mode)
+		chassis.leg_overload_mode = 1
+		chassis.step_in = min(1, round(chassis.step_in/2))
+		chassis.step_energy_drain = max(chassis.overload_step_energy_drain_min,chassis.step_energy_drain*chassis.leg_overload_coeff)
+		chassis.occupant_message("<span class='danger'>You enable leg actuators overload.</span>")
+	else
+		chassis.leg_overload_mode = 0
+		chassis.step_in = initial(chassis.step_in)
+		chassis.step_energy_drain = chassis.normal_step_energy_drain
+		chassis.occupant_message("<span class='notice'>You disable leg actuators overload.</span>")
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
	chassis.step_in = chassis.leg_overload_mode ? min(1, round(chassis.step_in/2)) : initial(chassis.step_in)
	chassis.step_energy_drain = chassis.leg_overload_mode ? max(chassis.overload_step_energy_drain_min,chassis.step_energy_drain*chassis.leg_overload_coeff) : chassis.normal_step_energy_drain
	chassis.occupant_message("<span class='danger'>You [chassis.leg_overload_mode ? "enable" : "disable"] leg actuators overload.</span>")
```

Comment (bobbahbrown) [reply code/game/mecha/combat/nerchen.dm]: idk why you were setting chassis.leg_overload_mode in that way it did nothing

**Review by bobbah — CHANGES_REQUESTED** [review]

File: code/game/mecha/combat/nerchen.dm (hunk 9 above)
Comment (tralezab) [reply code/game/mecha/combat/nerchen.dm]: sorry vestigial code

File: code/game/mecha/combat/nerchen.dm (hunk 8 above)
Comment (tralezab) [reply code/game/mecha/combat/nerchen.dm]: see below

File: code/game/mecha/combat/nerchen.dm (hunk 7 above)
Comment (tralezab) [reply code/game/mecha/combat/nerchen.dm]: will resolve when i do it

**Comment by tralezab** [comment]
thanks bob you'rte the best

File: code/game/mecha/combat/nerchen.dm [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+	. = "-"
+	if(occupant)
+		. += "closed"
+	else
+		. += "open"
+	. += "-"
+	if(chen.occupant)
+		. += "closed"
+	else
+		. += "open"
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
	return "-[occupant ? "closed" : "open"]-[chen.occupant ? "closed" : "open"]"
```

File: code/game/mecha/combat/nerchen.dm [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+//leaping interactions with living mobs and mechas
+/obj/mecha/combat/nerchen/throw_impact(atom/hit_atom, datum/thrownthing/throwingdatum)
+	if(!leaping)
+		return ..()
+	leaping = FALSE
+	rebuild_icon()
+	if(ismecha(hit_atom) || isliving(hit_atom))
+		var/atom/movable/AM = hit_atom
+		var/turf/throwbackwards = get_step(src, turn(dir, 180)) //get the turf behind the mecha
+		visible_message("<span class='danger'>[src] throws [AM] over their head, smashing them into [throwbackwards]!")
```
Comment (bobbahbrown) [inline code/game/mecha/combat/nerchen.dm]:
```suggestion
		visible_message("<span class='danger'>[src] throws [AM] over their head, smashing them into [throwbackwards]!</span>")
```

**Comment by Pumpkinoe** [comment]
>:D good

**Comment by HomocidlePancakes** [comment]
Based.

**Comment by HomocidlePancakes** [comment]
Based.

**Comment by Tharcoonvagh** [comment]
PICTURES

**Comment by cacogen** [comment]
daddys back you bitches

**Comment by yeeyeh** [comment]
hi yes i am arm's sprite slave here is what it looks like
image

**Comment by LordSaladin** [comment]
is this a  specific rim refrence?

**Comment by tralezab** [comment]
it dont have to be

**Comment by Shadowflame909** [comment]
Can we have the mechs split off into highly weaker animal forms capable of matching a simple mob in a 1v1

Go go power rangers

**Comment by optimumtact** [comment]
Mechs should be updated to support multi seating, this hack of embedding a mecha inside a mecha is bad and frankly if any maintainer merges this I will make them fix any issues for the rest of their life.

**Comment by tralezab** [comment]
I've been offered a lot of possible solutions and I'll pick one to fix this issue

**Comment by Coastfront** [comment]
Sounds like bloodbrother heaven

File: code/datums/action.dm [hunk 12]
```diff
@@ -625,7 +625,7 @@
 /datum/action/cooldown
 	check_flags = NONE
 	transparent_when_unavailable = FALSE
-	var/cooldown_time = 0
+	var/cooldown_time = 0 //use this for the cooldown
```
Comment (Rohesie) [inline code/datums/action.dm:628]: Properly document the variable.

File: code/game/mecha/combat/nerchen.dm [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+	max_temperature = 25000
+	infra_luminosity = 3
+	max_equip = 0
+	wreckage = /obj/structure/mecha_wreckage/nerchen
+	melee_can_hit = FALSE
+	add_req_access = 1
+	internal_damage_threshold = 25
+	var/leaping = FALSE
+	var/obj/mecha/combat/chen/chen //keeps the other internal mech in check
+	var/datum/action/innate/mecha/leap/leap_action = new
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:20]: Document the variables.

File: code/game/mecha/combat/nerchen.dm [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+
+/obj/mecha/combat/nerchen/Initialize()
+	. = ..()
+	chen = new(src, src)
+	rebuild_icon()
+
+/obj/mecha/combat/nerchen/GrantActions(mob/living/user, human_occupant = 0)
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:28]:
```suggestion
	. = ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+	leap_action.Grant(user, src)
+
+/obj/mecha/combat/nerchen/RemoveActions(mob/living/user, human_occupant = 0)
+	..()
+	leap_action.Remove(user)
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:33]:
```suggestion
	leap_action.Remove(user)
	return ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+
+
+/obj/mecha/combat/nerchen/obj_destruction()
+	chen.Destroy()
+	..()
+/obj/mecha/combat/nerchen/Destroy()
+	chen.Destroy()
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:41]: Don't call `Destroy()` outside of qdel

File: code/game/mecha/combat/nerchen.dm [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+
+/obj/mecha/combat/nerchen/GrantActions(mob/living/user, human_occupant = 0)
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:44]:
```suggestion
	. = ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+	defense_action.Grant(user, src)
+
+/obj/mecha/combat/nerchen/RemoveActions(mob/living/user, human_occupant = 0)
+	..()
+	defense_action.Remove(user)
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:49]:
```suggestion
	defense_action.Remove(user)
	return ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+			var/mob/living/L = AM
+			L.Knockdown(5 SECONDS)
+			L.adjustBruteLoss(10) //not meant as a huge damage dealer
+		else
+			var/obj/mecha/mech = AM //but for mechs it is
+			mech.take_damage(50)
+			mech.emp_act(EMP_LIGHT)
+			for(var/i in 1 to 3)
+				mech.spark_system.start()
+				sleep(2)
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:78]: Use `spark_system.set_up` instead of doing a sleep-for

File: code/game/mecha/combat/nerchen.dm [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+
+/obj/mecha/combat/nerchen/proc/build_cockpit_state() //checks which cockpits are open
+	return "-[occupant ? "closed" : "open"]-[chen.occupant ? "closed" : "open"]"
+
+/obj/mecha/combat/nerchen/moved_inside(mob/living/carbon/human/H)
+	. = ..()
+	rebuild_icon()
+
+/obj/mecha/combat/nerchen/go_out(forced, atom/newloc = loc)
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:96]:
```suggestion
	. = ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+		return FALSE
+
+/obj/mecha/combat/nerchen/MouseDrop_T(mob/M, mob/user) //if chen exists, you can enter that instead.
+	switch(alert("Which cockpit would you like to enter?","Mecha","Ner (Movement)","Chen (Weapons)", "Cancel"))
+		if("Chen (Weapons)")
+			chen.MouseDrop_T(M, user)
+			return
+		if("Cancel")
+			return
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:111]:
```suggestion
	return ..()
```

File: code/game/mecha/combat/nerchen.dm [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+/obj/mecha/combat/chen/moved_inside(mob/living/carbon/human/H)
+	. = ..()
+	ner.rebuild_icon()
+
+/obj/mecha/combat/chen/domove(direction)
+	to_chat(occupant, "<span class='warning'>You need to be in the other cockpit to move!</span>")
+	return FALSE
+
+/obj/mecha/combat/chen/go_out(forced, atom/newloc = ner.loc)
+	..()
```
Comment (Rohesie) [inline code/game/mecha/combat/nerchen.dm:200]:
```suggestion
	. = ..()
```

File: code/game/mecha/mecha_construction_paths.dm [hunk 23]
```diff
@@ -957,6 +957,148 @@
...
+	circuit_weapon = /obj/item/circuitboard/mecha/nerchen/targeting
+
+	inner_plating = /obj/item/stack/sheet/plasteel
+	inner_plating_amount = 10
+
+	outer_plating = /obj/item/mecha_parts/part/nerchen_armor
+	outer_plating_amount = 1
+
+/datum/component/construction/mecha/nerchen/custom_action(obj/item/I, mob/living/user, diff)
+	if(!..())
```
Comment (Rohesie) [inline code/game/mecha/mecha_construction_paths.dm:989]:
```suggestion
	. = ..()
	if(!.)
```

**Comment by ExcessiveUseOfCobblestone** [comment]
The PR contents is going to change significantly given the above comments so Re-PR when you've updated it.

**Pull request closed without merging**